feat: Merchant can customize Product Carousels with inner components ex. Add To Cart

justinlee01 opened the pull request:
Closes [CXSPA-191](https://jira.tools.sap/browse/CXSPA-191)

- Reusing item template code in product-carousel.component.html into its own component (ProductCarouselItemComponent)
- Creating a stylesheet for the above component
- The above component accepts inner components. to be used for Add to Cart or Configure components

ArkadiySukharenko committed:
Add add-to-cart to product carousel

ArkadiySukharenko committed:
Comment out a ProductAddToCartComponent to hide it by default

justinlee01 committed:
Fixing formatting issues with Prettier

justinlee01 committed:
Cannot use absolute path outside of spec files - exporting ProductCarouselComponent sub-component

justinlee01 commented on projects/storefrontlib/cms-components/product/carousel/product-carousel-item/product-carousel-item.component.ts:
```diff
@@ -0,0 +1,30 @@
+  ],
+})
+export class ProductCarouselItemComponent implements OnChanges {
+  @Input() item: any;
```
```suggestion
  @Input() item: Product;
```

justinlee01 commented on projects/storefrontlib/cms-components/product/carousel/index.ts:
```diff
@@ -1,6 +1,7 @@
 export * from './product-carousel.model';
 export * from './product-carousel.service';
 export * from './product-carousel/product-carousel.component';
+export * from './product-carousel-item/product-carousel-item.component';
```
```suggestion
```

justinlee01 commented on projects/storefrontlib/cms-components/product/carousel/product-carousel/product-carousel.module.ts:
```diff
@@ -21,17 +23,25 @@ import { ProductCarouselComponent } from './product-carousel.component';
+                // 'ProductAddToCartComponent',
+              ],
+            },
+          },
```
```suggestion
```

justinlee01 commented on projects/storefrontlib/cms-components/product/carousel/product-carousel/product-carousel.component.spec.ts:
```diff
@@ -113,6 +114,7 @@ describe('ProductCarouselComponent', () => {
         imports: [RouterTestingModule, I18nTestingModule],
         declarations: [
           ProductCarouselComponent,
+          ProductCarouselItemComponent,
```
Mock this component and move element-specific tests to product-carousel-item.component.spec.ts

justinlee01 committed:
Using mock ProductCarouselItem for product-carousel tests and moving UI tests to product-carousel-item
Providing type for Product
Removing unnecessary config in product-carousel module as it goes unseen by the customer anyway
SCSS for showing multiple CMS components in product-carousel-item

justinlee01 committed:
Result from prettier:fix

justinlee01 committed:
E2E test for customizing Product Carousel component with Add to Cart button
Also shortening expression

justinlee01 committed:
Merge branch 'develop' into HEAD

justinlee01 committed:
Result from prettier:fix

plabadie commented on projects/core/src/config/test-config.module.ts:
```diff
@@ -58,7 +58,7 @@ export class TestConfigModule {
         {
           provide: TEST_CONFIG_COOKIE_NAME,
-          useValue: options && options.cookie,
+          useValue: options?.cookie,
```
Did this cause an issue?

justinlee01 replied:
No issue I perceived - this should be _more or less_ equivalent.

If `options` is the value `0` or `false`, then the code is _not_ equivalent. If that is the case, then I can revert this change.

plabadie replied:
If there is no added value to change this, then perhaps we don't.  Less noise in the PR and one less thing that can cause a surprise down the road.

plabadie commented on projects/storefrontlib/cms-components/product/carousel/product-carousel/product-carousel.module.ts:
```diff
@@ -21,6 +23,7 @@ import { ProductCarouselComponent } from './product-carousel.component';
     RouterModule,
     UrlModule,
     I18nModule,
+    PageComponentModule,
```
What is this used for?

justinlee01 replied:
This exports `InnerComponentsHostDirective`, which is used in the declaration of `ProductCarouselItemDirective`.

plabadie replied:
`InnerComponentsHostDirective`, thank!

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/product-carousel/product-carousel.e2e-spec.ts:
```diff
@@ -0,0 +1,39 @@
+      });
+    });
+  });
+});
```
Ideally, we should test this on the CI before merging and the non-core tests don't run for feature branches.
One way I can think of is temporarily renaming the test file to `.core-e2e-spec.ts` so it runs for the feature branch PR.  Then rename it back after assessing the success of the test.  On the long run, we should find an easier way though.

justinlee01 replied:
Sure, I can do that.

justinlee01 committed:
product-carousel.e2e-spec.ts -> product-carousel.core-e2e-spec.ts

plabadie commented on projects/storefrontlib/cms-components/product/carousel/product-carousel-item/product-carousel-item.component.ts:
```diff
@@ -0,0 +1,31 @@
+      provide: ProductListItemContext,
+      useExisting: ProductListItemContextSource,
+    },
+  ],
+})
```
As a convention, Spartacus uses `changeDetection: ChangeDetectionStrategy.OnPush,` on components.

plabadie commented on projects/storefrontlib/cms-components/product/carousel/product-carousel/product-carousel.module.ts:
```diff
@@ -31,7 +34,7 @@ import { ProductCarouselComponent } from './product-carousel.component';
-  declarations: [ProductCarouselComponent],
-  exports: [ProductCarouselComponent],
+  declarations: [ProductCarouselComponent, ProductCarouselItemComponent],
+  exports: [ProductCarouselComponent, ProductCarouselItemComponent],
```
Perhaps we create a module for `ProductCarouselItemComponent`.  What do you think?

justinlee01 replied:
I'm wondering if that's unnecessary - the `ProductListModule` declares `ProductListItemComponent`, etc.
Would the benefit be for lazy loading?

plabadie replied:
Either way is fine.  I was trying to look through the other components for what seems more consistent.  But if the product list and the product list item are in the same module, it's fine to leave it as is.

plabadie commented on projects/storefrontlib/cms-components/product/carousel/product-carousel/product-carousel.component.html:
```diff
@@ -11,17 +11,5 @@
-      {{ item.price?.formattedValue }}
-    </div>
-  </a>
+  <cx-product-carousel-item [item]="item"></cx-product-carousel-item>
```
We need some migration documentation for this slight dom change.
See docs/migration/5_0.md for some template change doc.

justinlee01 committed:
Setting the ChangeDetectionStrategy for ProductCarouselItemComponent to OnPush

justinlee01 committed:
Adding documentation for breaking change for ProductCarouselComponent

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/product-carousel/product-carousel.core-e2e-spec.ts:
```diff
@@ -0,0 +1,39 @@
+import * as cart from '../../../helpers/cart';
```
I suggest not making this a "core" e2e test. 

You probably changed it to core to test it out on the CI, but don't forget to change it back to "non-core".

plabadie commented on projects/storefrontlib/cms-components/product/carousel/product-carousel-item/product-carousel-item.component.spec.ts:
```diff
@@ -0,0 +1,278 @@
+        ],
+      })
+        .overrideComponent(ProductCarouselItemComponent, {
+          set: { changeDetection: ChangeDetectionStrategy.Default },
```
Was it required to override the change detection strategy for the unit tests to work?
Otherwise it's best t test the component as-is.

justinlee0777 replied:
No, I think I can omit this.

plabadie commented:
As a convention, the branch name should match the ticket ID, like `feature/CXSPA-191`.

plabadie commented on docs/migration/5_0.md:
```diff
@@ -654,3 +654,7 @@ Due to i18next migration, certain set of keys have been migrated (from `_plural`
+
+### ProductCarouselComponent
+
+- Template code for carousel item has been moved to a reusable component - `ProductCarouselItemComponent`.
```
Can you please mention the new tag `<cx-product-carousel-item>` that will appear in the DOM.  Thanks!

plabadie commented on projects/storefrontlib/cms-components/product/carousel/product-carousel-item/product-carousel-item.component.spec.ts:
```diff
@@ -0,0 +1,278 @@
+  });
+});
+
+describe('ProductCarouselItemComponent in product-carousel with add-to-cart component', () => {
```
This part I think is more of an integration test (covered by the e2e) and doesn't belong in the unit tests.
As far as  ProductCarouselItemComponent is concerned, the only thing it's responsible for is to have  cxInnerComponentsHost in its template  and that on change, it emits a new item value in the context.

justinlee0777 replied:
I shall delete this test then - the above case is covered in lines 148 - 158 already.

justinlee01 committed:
Renaming core E2E file to non-core
Removing unnecessary unit test for CarouselItemComponent
ChangeDetectionStrategy does not need to be overriden for CarouselItemComponent unit tests
Undoing unnecessary change for TestConfigModule
Updating migration docs to have selector of new CarouselItemComponent
